### PR TITLE
(CDAP-13598) Synchronize correctly with the class loader lock

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

File: cdap-spark-core-base/src/test/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoaderTest.java [hunk 1]
```diff
@@ -77,4 +82,42 @@ public void run() {
       }
     }
   }
+
+  @Test
+  public void testConcurrentLoadClass() throws Exception {
```
Comment (anew) [inline cdap-spark-core-base/src/test/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoaderTest.java:87]: thanks for writing a test.

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoader.java [hunk 2]
```diff
@@ -107,33 +107,35 @@ public URL apply(String resourceName) {
...
-      }
-
-      byte[] byteCode = rewriter.rewriteClass(name, is);
-
-      // If no rewrite was performed, just define the class with this classloader by calling findClass.
-      if (byteCode == null) {
-        cls = findClass(name);
-      } else {
-        cls = defineClass(name, byteCode, 0, byteCode.length);
+    synchronized (getClassLoadingLock(name)) {
```
Comment (anew) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoader.java:110]: is there a performance implication of synchronizing on every class load? Could the findLoadedClass() be outside? Would it help much?

**Review by anew — COMMENTED** [review]
Overall looks good, please see my performance question

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoader.java (hunk 2 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoader.java:110]: It is what's being done in the default java ClassLoader implementation. I don't think that's a big concern here.

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT6570

**Comment by chtyim** [comment]
Build passed.

**Pull request merged**